registry/storage/driver/filesystem/driver.go

```diff
@@ -291,6 +292,58 @@ func (d *driver) Move(ctx context.Context, sourcePath string, destPath string) e
 	}
 
 	err := os.Rename(source, dest)
+	if err != nil {
+		// Copy then delete fall back
+		linkErr := err.(*os.LinkError)
+		if errno, ok := linkErr.Err.(syscall.Errno); ok && errno == syscall.EXDEV || os.IsExist(err) {
+			err = d.copyFile(source, dest)
+			if err = os.RemoveAll(source); err != nil {
+				return err
+			}
+		}
+	}
+	return err
+}
+
+// Copy a file to another. If dst file exists, it will be overwritten
+func (d *driver) copyFile(src, dst string) (err error) {
+	in, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	// Allocate tmp file with unique name, otherwise multiple processes may write to the same file
+	out, err := ioutil.TempFile(path.Dir(dst), path.Base(dst))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// In any case, cleanup
+		if cierr := in.Close(); err == nil {
+			err = cierr
+		}
+		if coerr := out.Close(); err == nil {
+			err = coerr
+		}
+		if rmerr := os.RemoveAll(out.Name()); err == nil {
+			err = rmerr
+		}
+	}()
+	if _, err = io.Copy(out, in); err != nil {
+		return err
+	}
+	if err := out.Sync(); err != nil {
+		return err
+	}
+	mverr := os.Rename(out.Name(), dst)
+	defer func() {
```

Comment:
I don't see why this defer is needed, there is nothing called after this

Reply:
If I remember well, I added this defer so it overrides the errors from the first defer when rename passed. 

That was following one of your previous comments :
> in fact if rename passed the close should still happen but probably shouldn't return an error (after rename passed this function succeeded and returning an error might be unexpected to the caller)